In [1]:
import pandas as pd

In [2]:
root_path = './dataset/'

In [3]:
comportamiento_df = pd.read_csv(f'{root_path}comportamiento_tarjetasvisa.csv', sep = ',')

In [4]:
informacion_df = pd.read_csv(f'{root_path}informacion_adicional_tarjetas.csv', sep = ',')

In [5]:
comportamiento_df['ID_CLIENTE'].value_counts().shape[0]

45985

### Definición de mal pagador

In [6]:
comportamiento_df.drop_duplicates()
comportamiento_df

ID_CLIENTE  MESES  DIAS_VENCIDOS
0           5001711      0            NaN
1           5001711     -1           11.0
2           5001711     -2           16.0
3           5001711     -3           16.0
4           5001712      0            0.0
...             ...    ...            ...
1048570     5150487    -25            0.0
1048571     5150487    -26            0.0
1048572     5150487    -27            0.0
1048573     5150487    -28            0.0
1048574     5150487    -29            0.0

[1048575 rows x 3 columns]

In [7]:
comportamiento_df.isnull().sum()

ID_CLIENTE            0
MESES                 0
DIAS_VENCIDOS    209230
dtype: int64

##### 20% de las observaciones son nulas en columna DIAS_VENCIDOS

In [8]:
grupo_df = comportamiento_df.groupby(["ID_CLIENTE","MESES"]).max()
grupo_df

DIAS_VENCIDOS
ID_CLIENTE MESES               
5001711    -3              16.0
           -2              16.0
           -1              11.0
            0               NaN
5001712    -18             17.0
...                         ...
5150487    -4               0.0
           -3               0.0
           -2               0.0
           -1               0.0
            0               0.0

[1048575 rows x 1 columns]

In [9]:
filtro_nulos = comportamiento_df.groupby('ID_CLIENTE').filter(lambda x:x['DIAS_VENCIDOS'].isnull().any())
filtro_nulos['ID_CLIENTE'].value_counts()

5059942    61
5005201    61
5002806    61
5105565    61
5017058    61
           ..
5047681     1
5021959     1
5125464     1
5092092     1
5137027     1
Name: ID_CLIENTE, Length: 25005, dtype: int64

In [10]:
comportamiento_df['ID_CLIENTE'].value_counts()

5016769    61
5002806    61
5118192    61
5145767    61
5078567    61
           ..
5028795     1
5148442     1
5053900     1
5046332     1
5079047     1
Name: ID_CLIENTE, Length: 45985, dtype: int64

##### De un total de 45,985 clientes con comportamiento del manejo de tarjeta, 25,005 clientes tienen al menos un campo nulo en dias vencidos.

##### Se consideran todas las observaciones para definir el tipo de pagador toda vez que podemos asegurar un mal comportamiento con información cierta.

##### Score de comportamiento:
Mayor que 25 y menor igual que 30, se puntúa el comportamiento con 0.25.
Mayor que 30, se puntúa con 1.

In [11]:
# funcion para score comportamiento
def score_comportamiento(dias_vencidos):
    if dias_vencidos > 25 and dias_vencidos <= 30:
        return 0.25
    elif dias_vencidos > 30:
        return 1
    else:
        return 0

In [12]:
comportamiento_df['SCORE'] = comportamiento_df['DIAS_VENCIDOS'].apply(score_comportamiento)
comportamiento_df['MESES_ANTIGUEDAD'] = (comportamiento_df.groupby('ID_CLIENTE')['MESES'].transform('min')) * (-1)
comportamiento_df

ID_CLIENTE  MESES  DIAS_VENCIDOS  SCORE  MESES_ANTIGUEDAD
0           5001711      0            NaN    0.0                 3
1           5001711     -1           11.0    0.0                 3
2           5001711     -2           16.0    0.0                 3
3           5001711     -3           16.0    0.0                 3
4           5001712      0            0.0    0.0                18
...             ...    ...            ...    ...               ...
1048570     5150487    -25            0.0    0.0                29
1048571     5150487    -26            0.0    0.0                29
1048572     5150487    -27            0.0    0.0                29
1048573     5150487    -28            0.0    0.0                29
1048574     5150487    -29            0.0    0.0                29

[1048575 rows x 5 columns]

In [13]:
calificacion_df = comportamiento_df[["ID_CLIENTE","MESES_ANTIGUEDAD","SCORE"]].groupby(["ID_CLIENTE","MESES_ANTIGUEDAD"]).sum().reset_index()
calificacion_df

ID_CLIENTE  MESES_ANTIGUEDAD  SCORE
0         5001711                 3   0.00
1         5001712                18   0.25
2         5001713                21   0.00
3         5001714                14   0.00
4         5001715                59   0.00
...           ...               ...    ...
45980     5150482                28   0.25
45981     5150483                17   0.00
45982     5150484                12   0.50
45983     5150485                 1   0.25
45984     5150487                29   0.00

[45985 rows x 3 columns]

In [14]:
# definir etiqueta de mal pagador
def tipo_cliente(meses_antiguedad, score_total):
    if meses_antiguedad <= 6 and score_total > 0:
        return 'mal pagador'
    elif meses_antiguedad > 6 and meses_antiguedad <= 12 and score_total > 0.5:
        return 'mal pagador'
    elif meses_antiguedad > 12 and meses_antiguedad <= 24 and score_total >= 1.5:
        return 'mal pagador'
    elif meses_antiguedad > 24 and meses_antiguedad <= 36 and score_total >= 2.5:
        return 'mal pagador'
    elif meses_antiguedad > 36 and meses_antiguedad <= 48 and score_total >= 3.5:
        return 'mal pagador'
    elif meses_antiguedad > 48 and meses_antiguedad <= 60 and score_total >= 4.5:
        return 'mal pagador'
    elif meses_antiguedad > 60 and score_total >= 5.5:
        return 'mal pagador'
    else:
        return 'buen pagador'

In [15]:
calificacion_df['TIPO_CLIENTE'] = calificacion_df.apply(lambda x: tipo_cliente(x['MESES_ANTIGUEDAD'], x['SCORE']), axis=1)
calificacion_df['TIPO_CLIENTE'].value_counts()

buen pagador    41728
mal pagador      4257
Name: TIPO_CLIENTE, dtype: int64

### Join con información adicional del cliente

In [16]:
calificacion_df.isnull().sum()

ID_CLIENTE          0
MESES_ANTIGUEDAD    0
SCORE               0
TIPO_CLIENTE        0
dtype: int64

In [17]:
calificacion_df[calificacion_df["TIPO_CLIENTE"].isnull() == True]

Empty DataFrame
Columns: [ID_CLIENTE, MESES_ANTIGUEDAD, SCORE, TIPO_CLIENTE]
Index: []

No se tiene valores nulos con el nuevo dataset con la definición de mal pagador

In [18]:
#Se crea un nuevo data con las dos columnas solicitadas CLIENTE_ID (ID_CLIENTE) y mal_pagador (TIPO_CLIENTE)
definicion_df = calificacion_df[["ID_CLIENTE", "TIPO_CLIENTE"]]
definicion_df

ID_CLIENTE  TIPO_CLIENTE
0         5001711  buen pagador
1         5001712  buen pagador
2         5001713  buen pagador
3         5001714  buen pagador
4         5001715  buen pagador
...           ...           ...
45980     5150482  buen pagador
45981     5150483  buen pagador
45982     5150484  buen pagador
45983     5150485   mal pagador
45984     5150487  buen pagador

[45985 rows x 2 columns]

In [19]:
informacion_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 438557 entries, 0 to 438556
Data columns (total 18 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   ID_CLIENTE                  438557 non-null  int64  
 1   GENERO                      438557 non-null  object 
 2   TIENE_CARRO                 438557 non-null  int64  
 3   TIENE_PROPIEDADES           438557 non-null  int64  
 4   N_NINOS                     438557 non-null  int64  
 5   INGRESO_ANUAL               438557 non-null  float64
 6   CATEGORIA_INGRESO           438557 non-null  object 
 7   NIVEL_EDUCACION             438557 non-null  object 
 8   ESTADO_CIVIL                438557 non-null  object 
 9   TIPO_CASA                   438557 non-null  object 
 10  DIAS_DESDE_NACIMIENTO       438557 non-null  int64  
 11  DIAS_TRABAJANDO             438557 non-null  int64  
 12  TIENE_CELULAR               438557 non-null  int64  
 13  TIENE_NUMEROTE

In [20]:
informacion_df.drop_duplicates()
informacion_df

ID_CLIENTE GENERO  TIENE_CARRO  TIENE_PROPIEDADES  N_NINOS  \
0          5008804      M            1                  1        0   
1          5008805      M            1                  1        0   
2          5008806      M            1                  1        0   
3          5008808      F            0                  1        0   
4          5008809      F            0                  1        0   
...            ...    ...          ...                ...      ...   
438552     6840104      M            0                  1        0   
438553     6840222      F            0                  0        0   
438554     6841878      F            0                  0        0   
438555     6842765      F            0                  1        0   
438556     6842885      F            0                  1        0   

        INGRESO_ANUAL     CATEGORIA_INGRESO                NIVEL_EDUCACION  \
0            427500.0               Working               Higher education   
1            427500.0               Working               Higher education   
2            112500.0               Working  Secondary / secondary special   
3            270000.0  Commercial associate  Secondary / secondary special   
4            270000.0  Commercial associate  Secondary / secondary special   
...               ...                   ...                            ...   
438552       135000.0             Pensioner  Secondary / secondary special   
438553       103500.0               Working  Secondary / secondary special   
438554        54000.0  Commercial associate               Higher education   
438555        72000.0             Pensioner  Secondary / secondary special   
438556       121500.0               Working  Secondary / secondary special   

                ESTADO_CIVIL          TIPO_CASA  DIAS_DESDE_NACIMIENTO  \
0             Civil marriage   Rented apartment                 -12005   
1             Civil marriage   Rented apartment                 -12005   
2                    Married  House / apartment                 -21474   
3       Single / not married  House / apartment                 -19110   
4       Single / not married  House / apartment                 -19110   
...                      ...                ...                    ...   
438552             Separated  House / apartment                 -22717   
438553  Single / not married  House / apartment                 -15939   
438554  Single / not married       With parents                  -8169   
438555               Married  House / apartment                 -21673   
438556               Married  House / apartment                 -18858   

        DIAS_TRABAJANDO  TIENE_CELULAR  TIENE_NUMEROTELEF_LABORAL  \
0                 -4542              1                          1   
1                 -4542              1                          1   
2                 -1134              1                          0   
3                 -3051              1                          0   
4                 -3051              1                          0   
...                 ...            ...                        ...   
438552           365243              1                          0   
438553            -3007              1                          0   
438554             -372              1                          1   
438555           365243              1                          0   
438556            -1201              1                          0   

        TIENE_NUMEROTELEF_PERSONAL  TIENE_EMAIL       PROFESION  \
0                                0            0             NaN   
1                                0            0             NaN   
2                                0            0  Security staff   
3                                1            1     Sales staff   
4                                1            1     Sales staff   
...                            ...          ...             ...   
438552                           0         

In [21]:
informacion_df.isnull().sum()

ID_CLIENTE                         0
GENERO                             0
TIENE_CARRO                        0
TIENE_PROPIEDADES                  0
N_NINOS                            0
INGRESO_ANUAL                      0
CATEGORIA_INGRESO                  0
NIVEL_EDUCACION                    0
ESTADO_CIVIL                       0
TIPO_CASA                          0
DIAS_DESDE_NACIMIENTO              0
DIAS_TRABAJANDO                    0
TIENE_CELULAR                      0
TIENE_NUMEROTELEF_LABORAL          0
TIENE_NUMEROTELEF_PERSONAL         0
TIENE_EMAIL                        0
PROFESION                     134203
N_MIEMBROSFAMILIA                  0
dtype: int64

##### El 30% de las observaciones para la información de las tarjetas es nula con detalles sobre la profesión que tienen

In [22]:
unificado_df = informacion_df.merge(definicion_df, how="left", on="ID_CLIENTE")
unificado_df

ID_CLIENTE GENERO  TIENE_CARRO  TIENE_PROPIEDADES  N_NINOS  \
0          5008804      M            1                  1        0   
1          5008805      M            1                  1        0   
2          5008806      M            1                  1        0   
3          5008808      F            0                  1        0   
4          5008809      F            0                  1        0   
...            ...    ...          ...                ...      ...   
438552     6840104      M            0                  1        0   
438553     6840222      F            0                  0        0   
438554     6841878      F            0                  0        0   
438555     6842765      F            0                  1        0   
438556     6842885      F            0                  1        0   

        INGRESO_ANUAL     CATEGORIA_INGRESO                NIVEL_EDUCACION  \
0            427500.0               Working               Higher education   
1            427500.0               Working               Higher education   
2            112500.0               Working  Secondary / secondary special   
3            270000.0  Commercial associate  Secondary / secondary special   
4            270000.0  Commercial associate  Secondary / secondary special   
...               ...                   ...                            ...   
438552       135000.0             Pensioner  Secondary / secondary special   
438553       103500.0               Working  Secondary / secondary special   
438554        54000.0  Commercial associate               Higher education   
438555        72000.0             Pensioner  Secondary / secondary special   
438556       121500.0               Working  Secondary / secondary special   

                ESTADO_CIVIL          TIPO_CASA  DIAS_DESDE_NACIMIENTO  \
0             Civil marriage   Rented apartment                 -12005   
1             Civil marriage   Rented apartment                 -12005   
2                    Married  House / apartment                 -21474   
3       Single / not married  House / apartment                 -19110   
4       Single / not married  House / apartment                 -19110   
...                      ...                ...                    ...   
438552             Separated  House / apartment                 -22717   
438553  Single / not married  House / apartment                 -15939   
438554  Single / not married       With parents                  -8169   
438555               Married  House / apartment                 -21673   
438556               Married  House / apartment                 -18858   

        DIAS_TRABAJANDO  TIENE_CELULAR  TIENE_NUMEROTELEF_LABORAL  \
0                 -4542              1                          1   
1                 -4542              1                          1   
2                 -1134              1                          0   
3                 -3051              1                          0   
4                 -3051              1                          0   
...                 ...            ...                        ...   
438552           365243              1                          0   
438553            -3007              1                          0   
438554             -372              1                          1   
438555           365243              1                          0   
438556            -1201              1                          0   

        TIENE_NUMEROTELEF_PERSONAL  TIENE_EMAIL       PROFESION  \
0                                0            0             NaN   
1                                0            0             NaN   
2                                0            0  Security staff   
3                                1            1     Sales staff   
4                                1            1     Sales staff   
...                            ...          ...             ...   
438552                           0         

In [23]:
unificado_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 438557 entries, 0 to 438556
Data columns (total 19 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   ID_CLIENTE                  438557 non-null  int64  
 1   GENERO                      438557 non-null  object 
 2   TIENE_CARRO                 438557 non-null  int64  
 3   TIENE_PROPIEDADES           438557 non-null  int64  
 4   N_NINOS                     438557 non-null  int64  
 5   INGRESO_ANUAL               438557 non-null  float64
 6   CATEGORIA_INGRESO           438557 non-null  object 
 7   NIVEL_EDUCACION             438557 non-null  object 
 8   ESTADO_CIVIL                438557 non-null  object 
 9   TIPO_CASA                   438557 non-null  object 
 10  DIAS_DESDE_NACIMIENTO       438557 non-null  int64  
 11  DIAS_TRABAJANDO             438557 non-null  int64  
 12  TIENE_CELULAR               438557 non-null  int64  
 13  TIENE_NUMEROTE

Hay 408100 registros aún con información nula porque los ids de clientes no estaban en la definición del tipo de cliente.
Es un categoría necesaria para analizar el tipo de cliente se va a trabajar con clientes que si estén categorizados

In [24]:
unificado_df[unificado_df.duplicated() == True]

Empty DataFrame
Columns: [ID_CLIENTE, GENERO, TIENE_CARRO, TIENE_PROPIEDADES, N_NINOS, INGRESO_ANUAL, CATEGORIA_INGRESO, NIVEL_EDUCACION, ESTADO_CIVIL, TIPO_CASA, DIAS_DESDE_NACIMIENTO, DIAS_TRABAJANDO, TIENE_CELULAR, TIENE_NUMEROTELEF_LABORAL, TIENE_NUMEROTELEF_PERSONAL, TIENE_EMAIL, PROFESION, N_MIEMBROSFAMILIA, TIPO_CLIENTE]
Index: []

No se tiene valores duplicados ya que desde el inicio se usó la función drop_duplicates para no trabajar con valores repetidos

In [25]:
clientes_df = unificado_df[unificado_df["TIPO_CLIENTE"].isnull() == False]
clientes_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 36457 entries, 0 to 434812
Data columns (total 19 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   ID_CLIENTE                  36457 non-null  int64  
 1   GENERO                      36457 non-null  object 
 2   TIENE_CARRO                 36457 non-null  int64  
 3   TIENE_PROPIEDADES           36457 non-null  int64  
 4   N_NINOS                     36457 non-null  int64  
 5   INGRESO_ANUAL               36457 non-null  float64
 6   CATEGORIA_INGRESO           36457 non-null  object 
 7   NIVEL_EDUCACION             36457 non-null  object 
 8   ESTADO_CIVIL                36457 non-null  object 
 9   TIPO_CASA                   36457 non-null  object 
 10  DIAS_DESDE_NACIMIENTO       36457 non-null  int64  
 11  DIAS_TRABAJANDO             36457 non-null  int64  
 12  TIENE_CELULAR               36457 non-null  int64  
 13  TIENE_NUMEROTELEF_LABORAL   36

In [26]:
clientes_df[["TIPO_CLIENTE", "ID_CLIENTE"]].groupby("TIPO_CLIENTE").count()

ID_CLIENTE
TIPO_CLIENTE            
buen pagador       32959
mal pagador         3498

### Data quality

El 68% de la información que se ha recolectado del cliente no ha indicado su profesión.

In [27]:
segun_genero_df = clientes_df[["GENERO", "PROFESION", "ID_CLIENTE"]].groupby("GENERO").count()
segun_genero_df

PROFESION  ID_CLIENTE
GENERO                       
F           15630       24430
M            9504       12027

In [28]:
segun_genero_df["SIN PROFESION"] = segun_genero_df["ID_CLIENTE"] - segun_genero_df["PROFESION"]
segun_genero_df["% SIN PROFESION"] = (segun_genero_df["SIN PROFESION"] * 100)/segun_genero_df["ID_CLIENTE"]
segun_genero_df

PROFESION  ID_CLIENTE  SIN PROFESION  % SIN PROFESION
GENERO                                                       
F           15630       24430           8800        36.021285
M            9504       12027           2523        20.977800

El 36 % de mujeres no ha indicado que profesión tienen y el 21% de los hombres prefirieron no decir a que se dedican